### PR TITLE
Aliases

### PR DESCRIPTION
**Pull request by EPashkin**
Generated WidgetPath needed gkoz/gtk-rs-glib#2 to build

### CODE REVIEW

**Comment by EPashkin** [comment]
Build error because [generate_constants](https://github.com/gkoz/gir/blob/master/src/codegen/sys/lib_.rs#L120-L142) uses `parameter_rust_type` that returns not ffi type names.
Before code looked
```
//pub const PANGO_GLYPH_EMPTY: Unknown rust type: "Glyph" = 268435455;
//pub const PANGO_GLYPH_INVALID_INPUT: Unknown rust type: "Glyph" = 4294967295;
//pub const PANGO_GLYPH_UNKNOWN_FLAG: Unknown rust type: "Glyph" = 268435456;
```
now 
```
pub const PANGO_GLYPH_EMPTY: Glyph = 268435455;
pub const PANGO_GLYPH_INVALID_INPUT: Glyph = 4294967295;
pub const PANGO_GLYPH_UNKNOWN_FLAG: Glyph = 268435456;
```
and must use `PangoGlyph`.
I can add hardcoded fix to `generate_constants` same as for `"&str"`

**Comment by EPashkin** [comment]
Tried to find alias for replacing `Quark` but don't find any good, so just change conversion for him to reflect changes in glib.

**Comment by gkoz** [comment]
Hmm, I wonder how a PR, that touches neither parser nor sys codegen, managed to break the sys tests. I'll take a better overall look tomorrow.

**Comment by EPashkin** [comment]
See https://github.com/gkoz/gir/pull/162#issuecomment-156713825.
PR just uncomment some constants.

**Comment by gkoz** [comment]
Ah, okay. Hardcoded is fine, the ffi_ng branch is supposed to replace that anyway.

**Comment by EPashkin** [comment]
Fixed build, ready for merge

File: src/analysis/functions.rs [hunk 1]
```diff
@@ -89,6 +95,10 @@ fn analyze_function(env: &Env, func: &library::Function, type_tid: library::Type
                 imports.add(s, func.version);
             }
         }
+        //TODO: change to check upcasts count
```
Comment (EPashkin) [inline src/analysis/functions.rs]: Forgot about this TODO, doing it now

**Comment by EPashkin** [comment]
Done

**Comment by EPashkin** [comment]
Next attempt for alias, not base on `GdkRectangle` 
Still in progress.
```
    #[cfg(gtk_3_12)]
    fn set_pointing_to(&self, rect: &gdk::Rectangle) {
        unsafe {
            ffi::gtk_popover_set_pointing_to(self.upcast().to_glib_none().0, rect);
        }
    }
```

**Comment by EPashkin** [comment]
Counting GdkRectangle as direct arise problem with
`gtk_image_new_from_surface`: it lose optional
```
    #[cfg(gtk_3_10)]
    pub fn new_from_surface(surface: &cairo::Surface) -> Image {
        unsafe {
            Widget::from_glib_none(ffi::gtk_image_new_from_surface(surface)).downcast_unchecked()
        }
    }
```
not sure what better to do:
a) change conversion back to pointer or scalar (for that needed write wrapper in cairo and reexport it in Gdk instead ffi)
b) write some fix for body and optional records.
c) create other conversion type `DirectPointer`, that same as direct but allow optional

**Comment by gkoz** [comment]
Suppose we had a
```rust
#[repr(C)]
struct Rectangle {
    x1: i32,
    y1: i32,
    x2: i32,
    x2: i32,
}
```
and faked conversions by using transmutes
```rust
impl<'a> ToGlibPtr<'a, *const GdkRectange> for &'a Rectangle {
    type Storage = &'a Rectangle;
    fn to_glib_none(&self) -> Stash<'a, &'a Rectangle> {
        Stash(transmute(*self), *self)
    }
}
```

**Comment by EPashkin** [comment]
So it act as ConversionType::Pointer?
This conversion will be added to cairo (as I know rust don't allow this in other crates)?

**Comment by EPashkin** [comment]
Ignore last comment.

**Comment by EPashkin** [comment]
So my reexport commit need changed?

**Comment by EPashkin** [comment]
Also needed implementation for `Uninitialized` as it needed for out parameter in `gtk_range_get_range_rect`

**Comment by gkoz** [comment]
BTW the hacky translations aren't strictly necessary, we could start with ones that actually copy the data but are less risky.

I'd leave `gdk` alone for now.

It could be, handling aliases is more of a distraction at this time. We can't make the generator perfect so some functions will need to be done manually. Perhaps it's time to start experimenting with how exactly that is going to work.

**Comment by gkoz** [comment]
> I'd leave `gdk` alone for now

 -- i.e. this `struct Rectangle` could even be defined in `gtk` first.

**Comment by EPashkin** [comment]
So functions will look like this? (It need some hardcoding in `rust_type` to remove `gdk::`)
```
    #[cfg(gtk_3_12)]
    fn set_pointing_to(&self, rect: &Rectangle) {
        unsafe {
            ffi::gtk_popover_set_pointing_to(self.upcast().to_glib_none().0, rect.to_glib_none().0);
        }
    }

    #[cfg(gtk_3_10)]
    pub fn new_from_surface(surface: Option<&cairo::Surface>) -> Image {
        unsafe {
Widget::from_glib_none(ffi::gtk_image_new_from_surface(surface.to_glib_none().0)).downcast_unchecked()
        }
    }
```

**Comment by gkoz** [comment]
Yes, that looks good.

Eventually might be a hardcoding issue *if* we try to make gdk and cairo share the same type.

**Comment by EPashkin** [comment]
As `ToGlibPtr` already present in nonobject glib we can just declare `RectangleInt` in `cairo` and reexport it (and not GdkRectangle) in Gdk.

**Comment by EPashkin** [comment]
By way `cairo_rectangle_int_t` not (x1,y1)-(x2,y2) and (x1,x2,width, height)

**Comment by gkoz** [comment]
Thanks for the correction, I was too lazy to look it up. The point is moving from `c_int` in the original struct to `i32`, notably a different approach from what gi-rust is aiming to do, so this choice might be questioned or turn out wrong later.

**Comment by EPashkin** [comment]
How about conversion type `DirectPointer`, this don't needed any change in libs and work directly with ffi types?

**Comment by gkoz** [comment]
I don't have good arguments against that right now. Just that having `i32` and `u32` almost everywhere but one or two structs with `c_int`s doesn't look very nice. Especially if some ABI later manages to make `c_int != i32`.

**Comment by EPashkin** [comment]
I'm afraid of slightly more than one - two: GLib.DateTime, GdkPoint,  GtkTargetEntry (it boxed with fields and without getters, not sure that it really needed get\set for using as it has good constructor), GtkBorder(also boxed, but with empty constructor, important fields and no get\set), others in https://github.com/gkoz/gir/issues/103#issuecomment-153145754.
So this issue requires a lot of thinking.

**Comment by gkoz** [comment]
`DateTime` should be converted into whatever Rust counterpart we choose (e.g. `chrono::DateTime`) and is opaque, `TargetEntry` carries a pointer so it's definitely not POD. `Point` yep. `Border` is a weird case but at least the fields unambiguously are `u16`.

**Comment by EPashkin** [comment]
Tried update macro `glib_boxed_wrapper` for boxed records with fields with `TargetEntry` as example (https://github.com/EPashkin/rust-gnome-glib/tree/glib_boxed_wrapper_with_fields).
@gkoz , how it look for you?
```
glib_wrapper! {
    pub struct TargetEntry(Boxed<ffi::GtkTargetEntry>);

    fields {
        get pointer (get_target) target: String,
        get direct (get_flags) flags: u32,
        get direct (get_info) info: u32,
    }

    match fn {
        copy => |ptr| ffi::gtk_target_entry_copy(ptr as *mut ffi::GtkTargetEntry),
        free => |ptr| ffi::gtk_target_entry_free(ptr),
    }
}

impl TargetEntry {
    pub fn new(target: &str, flags: u32, info: u32) -> TargetEntry {
        unsafe {
            from_glib_full(ffi::gtk_target_entry_new(target.to_glib_none().0, flags, info))
        }
    }

    pub fn test(&self){
        let target: String = self.get_target();
        let flags:u32 = self.get_flags();
        let info:u32 = self.get_info();
    }
}
```

**Comment by gkoz** [comment]
The most complicated yet essential goal would be supporting [GdkEvent](https://developer.gnome.org/gdk3/stable/gdk3-Event-Structures.html) records.

`Deref` might be a better match than `AsRef`, you won't even have to call `as_ref` then.

The macros don't seem to buy us much because of all the variations. For instance, getting a string out might work better (assuming there is an appropriate implementation) as

```rust
    fn target(&self) -> Cow<str> {
        from_glib_borrow(self.0.target)
    }
```
but encoding this in the macro syntax complicates things. Even the integer case looks simple enough thanks to `Deref` (unless I'm missing something)
```rust
    fn info(&self) -> u32 {
        self.0.info
    }
```

**Comment by EPashkin** [comment]
I don't even want think about `GdkEvent` (all unions will be problems)
`Deref` is much better.
Updated branch, not sure if setter done right way.
About custom functions, we can make macro that accept functions like this
```
glib_wrapper! {
    pub struct TargetEntry(Boxed<ffi::GtkTargetEntry>);

    impl {
        fn target(&self) -> Cow<str> {
            from_glib_borrow(self.0.target)
        }
        fn info(&self) -> u32 {
            self.0.info
        }
    }

    match fn {
        copy => |ptr| ffi::gtk_target_entry_copy(ptr as *mut ffi::GtkTargetEntry),
        free => |ptr| ffi::gtk_target_entry_free(ptr),
    }
}
```
May be it be better as it more customizable.

**Comment by EPashkin** [comment]
As for `GdkEvent` can we do this?
Declare rust enum `Event_` with supported cases (possibly with not full field list) and `Unsupported` case and `struct Event(Event_, gdk_ffi::GdkEvent, MemoryManager<gdk_ffi::GdkEvent>)`
On `from_glib_full` create `Event_` by copying supported fields or `Event_::Unsupported` if case not supported.
On `to_glib_none` copy supported fields back to `gdk_ffi::GdkEvent` and return it.

**Comment by EPashkin** [comment]
Not sure if any fields of received event need be changed by rust code.
If no thus fields then `to_glib_none` will be simply returning stored `gdk_ffi::GdkEvent`.

Also now `gdk_ffi::GdkEvent = c_void` it needed to change to `[u8;N]` where `N` is size of union.

**Comment by EPashkin** [comment]
* If no thus fields => If no such fields

**Comment by EPashkin** [comment]
About `https://github.com/gkoz/gir/pull/162#issuecomment-159004477`
I was wrong: for custom accessors we don't need macro as we can just define `impl TargetEntry` with these functions.

**Comment by gkoz** [comment]
An enum doesn't allow to put a specific event type in a signature.

One possible strategy is to make `struct Event(Boxed<GdkEvent>)`, that converts to both `*mut GdkEvent` and `*mut GdkEventAny`, and make the rest wrap `Event`, e.g. `struct EventButton(Event)` with `Deref`.

**Comment by EPashkin** [comment]
I talked about internal type that can be filled from `GdkEvent` and if needed copy changes back.
```
enum Event_{
   Key{
       is_release: bool,
       send_event: i8,
       time: u32,
       keyval: u32,
   },
   Button{
       type: BtnEventType,
       send_event: i8,
       time: u32,
       x: double,
       y: double,
       state: u32,
   },
...
   Unsupported,
}
```

**Comment by EPashkin** [comment]
Oh, sorry, you saying about callback already, then you right, I missed that callback need accept single type.

**Comment by EPashkin** [comment]
Rebased. Last commit only needed if `Rectangle` will placed in `gtk`

**Comment by gkoz** [comment]
Looks like all aliases are tricky so this is difficult to evaluate. `Rectangle` needs `caller-allocates` support we don't have yet.

**Comment by EPashkin** [comment]
I done caller support already, add PR after this.

**Comment by EPashkin** [comment]
Or can add it to this PR

**Comment by gkoz** [comment]
Fine, I'll merge this and we'll fix any problems as they come.

@homu r+

**Pull request merged**